### ISSUE TITLE
Sortable in modal - `keydown` event propagation issue

### ISSUE THREAD

**Issue by wolkyura** [clauderic/dnd-kit#1367]
When using `useSortable` in a modal, there is an issue with keyboard event propagation. Currently, keyboard events are not prevented from propagation, causing unintended behavior. The modal catches `keydown` event (`Escape`) before `dnd-kit` listeners do, and closes instead of canceling the drag and stopping the propagation.

####  Expected behavior:
I'd expect that when keyboard navigation is activated within the sortable items, keyboard events should be stopped from further propagation. This would ensure that the modal (or other components) does not receive and act upon keyboard events when it shouldn't.

#### Codesandbox example
https://codesandbox.io/p/sandbox/priceless-currying-8ytlml?file=%2Fsrc%2FSortable.js%3A48%2C1

#### Steps to reproduce
- Open modal
- Press `Tab` to focus on one of the items
- Press `Space` or `Enter` to activate keyboard listeners
- Press `Escape`

#### Notes
I tried to prevent event propagation before invoking `useSortable` listeners, but this breaks the cancel functionality. I discovered that these events are attached to the document, causing them to be fired at the very end, which looks like the main issue. To fix this, I created `MyKeyboardSensor` and overrode `attach()` and `detach()` methods to set events directly to the target, and added `event.stopPropagation()`, this fixed my issue. I'm curious is there any particular reason why events are set to the document and not to the target itself?

```js
class MyKeyboardSensor extends KeyboardSensor {
  private attach() {
    this.handleStart();

    this.windowListeners.add('resize', this.handleCancel);
    this.windowListeners.add('visibilitychange', this.handleCancel);

    this.props.event.target.addEventListener('keydown', this.handleKeyDown);
  }

  private detach() {
    this.props.event.target.removeEventListener('keydown', this.handleKeyDown);
    this.windowListeners.removeAll();
  }

  private handleCancel(event: Event) {
    const { onCancel } = this.props;

    event.stopPropagation();
    event.preventDefault();
    this.detach();
    onCancel();
  }
}
```

Also, I think listeners lack `blur` event handling, I'd expect that drag is canceled when this event is fired on the target element. Let me know if I should create a separate issue for this.

By the way, thanks for creating such an awesome library!

**Comment by theenoahmason** [comment]
Having the same issue with event propogation on modals, and @wolkyura's explanation makes a lot of sense I feel like this should be implemented in the lib.

I can also confirm that blurring does not drop or cancel the draggable item. I fixed this for keyboard users by adding `Escape` to the `cancel` keyboard codes, but this does not handle proper blurring.

```typescript
const keyboardSensor = useSensor(KeyboardSensor, {
	...
	keyboardCodes: {
		...
		cancel: ["Escape", "Tab"],
	},
});
```

Blur should probably be natively handled by the library, but one way of allowing users to do this is If `useSortable` exposed a cancel function, that we could simply pass to `onBlur` of the handle. That said, I can't imagine any situation where someone would want keyboard sorting to continue after a blur.

**Comment by theenoahmason** [comment]
@wolkyura Thanks for getting me pointed in the right direction. Here is what I have come up with. It handles your methodology while adhering to the KeyboardSensor's methodology as well. It also handles blur cancellation effectively.

@clauderic I have been messing with this for days and I think this combination is pretty solid and could be easily placed into core. I don't have time at this very second to make a PR but I would be happy to.

```typescript
import { KeyboardSensor, KeyboardSensorProps } from "@dnd-kit/core";

/**
 * Creates a new keyboard sensor, extending the default `KeyboardSensor` class.
 */
export default class SortableContainerKeyboardSensor extends KeyboardSensor {

	constructor(props: KeyboardSensorProps) {
		super(props);
		/**
		 * Bind the new methods to the instance.
		 */
		this.handleBlur = this.handleBlur.bind(this);
	}

	/**
	 * Override the inherited `attach` method to add our new event handlers.
	 */
	protected attach() {
		this.handleStart();

		this.windowListeners.add("resize", this.handleCancel);
		this.windowListeners.add("visibilitychange", this.handleCancel);

		setTimeout(() => {
			/**
			 * Updates this event to fire in the capture phase.
			 */
			this.listeners.add("keydown", this.handleKeyDown, { capture: true });
			/**
			 * Add a blur event listener in the capture phase.
			 */
			this.listeners.add("blur", this.handleBlur, { capture: true });
		});
	}

	/**
	 * When the keyboard moves the active item around, it fires
	 * blur events, however in those events the `relatedTarget`
	 * is null. Check for the existence of this key to determine
	 * if the event should fire a cancel or not.
	 */
	protected handleBlur(event: FocusEvent) {
		if (event.relatedTarget instanceof HTMLElement) {
			this.handleCancel(event);
		}
	}

	/**
	 * Override the inherited `handleEnd` method and stop propagation.
	 */
	protected handleEnd(event: Event) {
		const {onEnd} = this.props;

		event.preventDefault();
		event.stopPropagation();

		this.detach();
		onEnd();
	}

	/**
	 * Override the inherited `handleCancel` method and stop propagation.
	 */
	protected handleCancel(event: Event) {
		const {onCancel} = this.props;

		event.preventDefault();
		event.stopPropagation();

		this.detach();
		onCancel();
	}
}
```

In addition, I have declared the default KeyboardSensor class interface as an ambient declaration on the module. I mapped the methods a little better than the type that is being exported by the lib, marking the methods as protected instead of private so I could extend them. In general, the default sensors should probably use protected members and methods instead of private ones so users can override them without the headache if needed.

```typescript
import { Activators, KeyboardSensorOptions, KeyboardSensorProps, SensorInstance } from "@dnd-kit/core";

declare class KeyboardSensor implements SensorInstance {
	protected props: KeyboardSensorProps;
	autoScrollEnabled: boolean;
	protected referenceCoordinates: Coordinates | undefined;
	protected listeners: Listeners;
	protected windowListeners: Listeners;
	constructor(props: KeyboardSensorProps);
	protected attach();
	protected handleStart();
	protected handleKeyDown(event: Event);
	protected handleMove(event: Event, coordinates: Coordinates);
	protected handleEnd(event: Event);
	protected handleCancel(event: Event);
	protected detach();
	static activators: Activators<KeyboardSensorOptions>;
}
```

**Comment by clauderic** [comment]
This issue was opened against a previous version of dnd-kit. The library has since undergone a major rewrite.

It has been rebuilt from the ground up with support for **multiple frameworks** (React, Vue, Svelte, Solid, and vanilla JavaScript), improved performance, and a new API.

- **Documentation**: https://dndkit.com
- **Migration guide**: https://dndkit.com/react/migration
- **GitHub**: https://github.com/clauderic/dnd-kit

If this issue is still relevant to the latest version, please open a new issue.
